src/nanorc/node.py

```diff
@@ -497,25 +496,39 @@ def _on_enter_callback(self, event):
             }
             timeout_bar = progress.add_task("[yellow]timeout", total = timeout*10)
 
+            failed_ping_count = {app.name:0 for app in appset}
+            mode_fail = []
             for _ in range(timeout*10):
                 progress.update(timeout_bar, advance=1)
-
                 if len(appset)==0:
                     progress.update(total, completed = n_apps)
                     progress.update(timeout_bar, visible=False)
                     break
                 done = []
                 for child_node in appset:
                     if not child_node.included: continue
-
-                    if not child_node.sup.desc.proc.is_alive() or not child_node.sup.commander.ping():
+                    is_alive = child_node.sup.desc.proc.is_alive()
+                    if not is_alive:
                         failed.append(child_node.name)
+                        mode_fail.append('app died')
                         child_node.to_error(
                             command = command,
                         )
                         done += [child_node]
-                        break
-
+                        continue
+                    is_ping = child_node.sup.commander.ping()
+                    if not is_ping:
+                        failed_ping_count[child_node.name] += 1
+                        failed_ping_thres = 3
```

Comment:
Please move the definition out of the loop or in a data member of the class.

Reply:
done in next commit